### PR TITLE
hotfix: capture and return empty list if category is wrong

### PR DESCRIPTION
**Pull request by PaulYuuu**
If the category is wrong, `find_supported_devices` will raise KeyError.
Captured it and return an empty list directly since the match was not
successful.

Bugzilla ID: 1781036
Signed-off-by: Yihuang Yu <yihyu@redhat.com>

### CODE REVIEW

**Comment by ldoktor** [comment]
Well this is not sufficient. First it still crashes:

```
$ avocado --show all run --vt-extra-params enable_kvm=no cpu_model=cortex-a57 machine_type_extra_params='' --vt-machine-type aarch64 --vt-arch arm64-pci --vt-qemu-bin /usr/local/bin/qemu-system-aarch64 -- boot
...
avocado.test: Reproduced traceback from: /home/medic/Work/Projekty/avocado/avocado-vt/avocado_vt/test.py:439
avocado.test: Traceback (most recent call last):
avocado.test:   File "/home/medic/Work/Projekty/avocado/avocado-vt/virttest/error_context.py", line 135, in new_fn
avocado.test:     return fn(*args, **kwargs)
avocado.test:   File "/home/medic/Work/Projekty/avocado/avocado-vt/virttest/env_process.py", line 1037, in preprocess
avocado.test:     "CPU")
avocado.test:   File "/home/medic/Work/Projekty/avocado/avocado-vt/virttest/utils_qemu.py", line 106, in find_supported_devices
avocado.test:     for device in get_supported_devices_list(bin_path, category):
avocado.test:   File "/home/medic/Work/Projekty/avocado/avocado-vt/virttest/utils_qemu.py", line 92, in get_supported_devices_list
avocado.test:     return list(get_devices_info(bin_path, category).keys())
avocado.test:   File "/home/medic/Work/Projekty/avocado/avocado-vt/virttest/utils_qemu.py", line 75, in get_devices_info
avocado.test:     device_type = re.match(r"([A-Z]\S+) devices:", device_info).group(1)
avocado.test: AttributeError: 'NoneType' object has no attribute 'group'
```

and second even if that worked, it'd result in incorrect values being supplied on arm (or other machines without default cpu) as it'd simply report no sections and no devices (and no cpu types).

You can checkout how to execute other archs here: https://avocado-vt.readthedocs.io/en/latest/RunningInEmulationTCG.html

**Comment by PaulYuuu** [comment]
Okay, my fault. Now qemu which not supports hotpluggable cpu device and/or require machine type will skip that step.

**Comment by luckyh** [comment]
@PaulYuuu LGTM, please provide the test result, thanks!

**Comment by PaulYuuu** [comment]
##### Results of latest code:
```python
>>> from virttest.utils_qemu import *
>>> for qemu_bin in ["qemu-system-aarch64", "qemu-system-i386", "qemu-system-ppc64", "qemu-system-s390x", "qemu-system-x86_64"]:
...     print(qemu_bin, ":\n")
...     print(has_device_category(qemu_bin, "CPU"))
...     print(get_supported_devices_list(qemu_bin, "CPU"), "\n")
... 
qemu-system-aarch64 :
False
[] 

qemu-system-i386 :
True
['486-i386-cpu', '486-v1-i386-cpu', 'athlon-i386-cpu', 'athlon-v1-i386-cpu', 'base-i386-cpu', 'Broadwell-i386-cpu', 'Broadwell-IBRS-i386-cpu', 'Broadwell-noTSX-i386-cpu', 'Broadwell-noTSX-IBRS-i386-cpu', 'Broadwell-v1-i386-cpu', 'Broadwell-v2-i386-cpu', 'Broadwell-v3-i386-cpu', 'Broadwell-v4-i386-cpu', 'Cascadelake-Server-i386-cpu', 'Cascadelake-Server-noTSX-i386-cpu', 'Cascadelake-Server-v1-i386-cpu', 'Cascadelake-Server-v2-i386-cpu', 'Cascadelake-Server-v3-i386-cpu', 'Conroe-i386-cpu', 'Conroe-v1-i386-cpu', 'core2duo-i386-cpu', 'core2duo-v1-i386-cpu', 'coreduo-i386-cpu', 'coreduo-v1-i386-cpu', 'Denverton-i386-cpu', 'Denverton-v1-i386-cpu', 'Dhyana-i386-cpu', 'Dhyana-v1-i386-cpu', 'EPYC-i386-cpu', 'EPYC-IBPB-i386-cpu', 'EPYC-v1-i386-cpu', 'EPYC-v2-i386-cpu', 'Haswell-i386-cpu', 'Haswell-IBRS-i386-cpu', 'Haswell-noTSX-i386-cpu', 'Haswell-noTSX-IBRS-i386-cpu', 'Haswell-v1-i386-cpu', 'Haswell-v2-i386-cpu', 'Haswell-v3-i386-cpu', 'Haswell-v4-i386-cpu', 'host-i386-cpu', 'Icelake-Client-i386-cpu', 'Icelake-Client-noTSX-i386-cpu', 'Icelake-Client-v1-i386-cpu', 'Icelake-Client-v2-i386-cpu', 'Icelake-Server-i386-cpu', 'Icelake-Server-noTSX-i386-cpu', 'Icelake-Server-v1-i386-cpu', 'Icelake-Server-v2-i386-cpu', 'IvyBridge-i386-cpu', 'IvyBridge-IBRS-i386-cpu', 'IvyBridge-v1-i386-cpu', 'IvyBridge-v2-i386-cpu', 'KnightsMill-i386-cpu', 'KnightsMill-v1-i386-cpu', 'kvm32-i386-cpu', 'kvm32-v1-i386-cpu', 'kvm64-i386-cpu', 'kvm64-v1-i386-cpu', 'max-i386-cpu', 'n270-i386-cpu', 'n270-v1-i386-cpu', 'Nehalem-i386-cpu', 'Nehalem-IBRS-i386-cpu', 'Nehalem-v1-i386-cpu', 'Nehalem-v2-i386-cpu', 'Opteron_G1-i386-cpu', 'Opteron_G1-v1-i386-cpu', 'Opteron_G2-i386-cpu', 'Opteron_G2-v1-i386-cpu', 'Opteron_G3-i386-cpu', 'Opteron_G3-v1-i386-cpu', 'Opteron_G4-i386-cpu', 'Opteron_G4-v1-i386-cpu', 'Opteron_G5-i386-cpu', 'Opteron_G5-v1-i386-cpu', 'Penryn-i386-cpu', 'Penryn-v1-i386-cpu', 'pentium-i386-cpu', 'pentium-v1-i386-cpu', 'pentium2-i386-cpu', 'pentium2-v1-i386-cpu', 'pentium3-i386-cpu', 'pentium3-v1-i386-cpu', 'phenom-i386-cpu', 'phenom-v1-i386-cpu', 'qemu32-i386-cpu', 'qemu32-v1-i386-cpu', 'qemu64-i386-cpu', 'qemu64-v1-i386-cpu', 'SandyBridge-i386-cpu', 'SandyBridge-IBRS-i386-cpu', 'SandyBridge-v1-i386-cpu', 'SandyBridge-v2-i386-cpu', 'Skylake-Client-i386-cpu', 'Skylake-Client-IBRS-i386-cpu', 'Skylake-Client-noTSX-IBRS-i386-cpu', 'Skylake-Client-v1-i386-cpu', 'Skylake-Client-v2-i386-cpu', 'Skylake-Client-v3-i386-cpu', 'Skylake-Server-i386-cpu', 'Skylake-Server-IBRS-i386-cpu', 'Skylake-Server-noTSX-IBRS-i386-cpu', 'Skylake-Server-v1-i386-cpu', 'Skylake-Server-v2-i386-cpu', 'Skylake-Server-v3-i386-cpu', 'Snowridge-i386-cpu', 'Snowridge-v1-i386-cpu', 'Snowridge-v2-i386-cpu', 'Westmere-i386-cpu', 'Westmere-IBRS-i386-cpu', 'Westmere-v1-i386-cpu', 'Westmere-v2-i386-cpu'] 

qemu-system-ppc64 :
True
['970_v2.2-spapr-cpu-core', '970mp_v1.0-spapr-cpu-core', '970mp_v1.1-spapr-cpu-core', 'power5+_v2.1-spapr-cpu-core', 'power7+_v2.1-spapr-cpu-core', 'power7_v2.3-spapr-cpu-core', 'power8_v2.0-powernv-cpu-core', 'power8_v2.0-spapr-cpu-core', 'power8e_v2.1-powernv-cpu-core', 'power8e_v2.1-spapr-cpu-core', 'power8nvl_v1.0-powernv-cpu-core', 'power8nvl_v1.0-spapr-cpu-core', 'power9_v1.0-spapr-cpu-core', 'power9_v2.0-powernv-cpu-core', 'power9_v2.0-spapr-cpu-core'] 

qemu-system-s390x :
True
['gen15a-base-s390x-cpu', 'gen15a-s390x-cpu', 'gen15b-base-s390x-cpu', 'gen15b-s390x-cpu', 'max-s390x-cpu', 'qemu-s390x-cpu', 'z10BC-base-s390x-cpu', 'z10BC-s390x-cpu', 'z10BC.2-base-s390x-cpu', 'z10BC.2-s390x-cpu', 'z10EC-base-s390x-cpu', 'z10EC-s390x-cpu', 'z10EC.2-base-s390x-cpu', 'z10EC.2-s390x-cpu', 'z10EC.3-base-s390x-cpu', 'z10EC.3-s390x-cpu', 'z114-base-s390x-cpu', 'z114-s390x-cpu', 'z13-base-s390x-cpu', 'z13-s390x-cpu', 'z13.2-base-s390x-cpu', 'z13.2-s390x-cpu', 'z13s-base-s390x-cpu', 'z13s-s390x-cpu', 'z14-base-s390x-cpu', 'z14-s390x-cpu', 'z14.2-base-s390x-cpu', 'z14.2-s390x-cpu', 'z14ZR1-base-s390x-cpu', 'z14ZR1-s390x-cpu', 'z196-base-s390x-cpu', 'z196-s390x-cpu', 'z196.2-base-s390x-cpu', 'z196.2-s390x-cpu', 'z800-base-s390x-cpu', 'z800-s390x-cpu', 'z890-base-s390x-cpu', 'z890-s390x-cpu', 'z890.2-base-s390x-cpu', 'z890.2-s390x-cpu', 'z890.3-base-s390x-cpu', 'z890.3-s390x-cpu', 'z900-base-s390x-cpu', 'z900-s390x-cpu', 'z900.2-base-s390x-cpu', 'z900.2-s390x-cpu', 'z900.3-base-s390x-cpu', 'z900.3-s390x-cpu', 'z990-base-s390x-cpu', 'z990-s390x-cpu', 'z990.2-base-s390x-cpu', 'z990.2-s390x-cpu', 'z990.3-base-s390x-cpu', 'z990.3-s390x-cpu', 'z990.4-base-s390x-cpu', 'z990.4-s390x-cpu', 'z990.5-base-s390x-cpu', 'z990.5-s390x-cpu', 'z9BC-base-s390x-cpu', 'z9BC-s390x-cpu', 'z9BC.2-base-s390x-cpu', 'z9BC.2-s390x-cpu', 'z9EC-base-s390x-cpu', 'z9EC-s390x-cpu', 'z9EC.2-base-s390x-cpu', 'z9EC.2-s390x-cpu', 'z9EC.3-base-s390x-cpu', 'z9EC.3-s390x-cpu', 'zBC12-base-s390x-cpu', 'zBC12-s390x-cpu', 'zEC12-base-s390x-cpu', 'zEC12-s390x-cpu', 'zEC12.2-base-s390x-cpu', 'zEC12.2-s390x-cpu'] 

qemu-system-x86_64 :
True
['486-v1-x86_64-cpu', '486-x86_64-cpu', 'athlon-v1-x86_64-cpu', 'athlon-x86_64-cpu', 'base-x86_64-cpu', 'Broadwell-IBRS-x86_64-cpu', 'Broadwell-noTSX-IBRS-x86_64-cpu', 'Broadwell-noTSX-x86_64-cpu', 'Broadwell-v1-x86_64-cpu', 'Broadwell-v2-x86_64-cpu', 'Broadwell-v3-x86_64-cpu', 'Broadwell-v4-x86_64-cpu', 'Broadwell-x86_64-cpu', 'Cascadelake-Server-noTSX-x86_64-cpu', 'Cascadelake-Server-v1-x86_64-cpu', 'Cascadelake-Server-v2-x86_64-cpu', 'Cascadelake-Server-v3-x86_64-cpu', 'Cascadelake-Server-x86_64-cpu', 'Conroe-v1-x86_64-cpu', 'Conroe-x86_64-cpu', 'core2duo-v1-x86_64-cpu', 'core2duo-x86_64-cpu', 'coreduo-v1-x86_64-cpu', 'coreduo-x86_64-cpu', 'Denverton-v1-x86_64-cpu', 'Denverton-x86_64-cpu', 'Dhyana-v1-x86_64-cpu', 'Dhyana-x86_64-cpu', 'EPYC-IBPB-x86_64-cpu', 'EPYC-v1-x86_64-cpu', 'EPYC-v2-x86_64-cpu', 'EPYC-x86_64-cpu', 'Haswell-IBRS-x86_64-cpu', 'Haswell-noTSX-IBRS-x86_64-cpu', 'Haswell-noTSX-x86_64-cpu', 'Haswell-v1-x86_64-cpu', 'Haswell-v2-x86_64-cpu', 'Haswell-v3-x86_64-cpu', 'Haswell-v4-x86_64-cpu', 'Haswell-x86_64-cpu', 'host-x86_64-cpu', 'Icelake-Client-noTSX-x86_64-cpu', 'Icelake-Client-v1-x86_64-cpu', 'Icelake-Client-v2-x86_64-cpu', 'Icelake-Client-x86_64-cpu', 'Icelake-Server-noTSX-x86_64-cpu', 'Icelake-Server-v1-x86_64-cpu', 'Icelake-Server-v2-x86_64-cpu', 'Icelake-Server-x86_64-cpu', 'IvyBridge-IBRS-x86_64-cpu', 'IvyBridge-v1-x86_64-cpu', 'IvyBridge-v2-x86_64-cpu', 'IvyBridge-x86_64-cpu', 'KnightsMill-v1-x86_64-cpu', 'KnightsMill-x86_64-cpu', 'kvm32-v1-x86_64-cpu', 'kvm32-x86_64-cpu', 'kvm64-v1-x86_64-cpu', 'kvm64-x86_64-cpu', 'max-x86_64-cpu', 'n270-v1-x86_64-cpu', 'n270-x86_64-cpu', 'Nehalem-IBRS-x86_64-cpu', 'Nehalem-v1-x86_64-cpu', 'Nehalem-v2-x86_64-cpu', 'Nehalem-x86_64-cpu', 'Opteron_G1-v1-x86_64-cpu', 'Opteron_G1-x86_64-cpu', 'Opteron_G2-v1-x86_64-cpu', 'Opteron_G2-x86_64-cpu', 'Opteron_G3-v1-x86_64-cpu', 'Opteron_G3-x86_64-cpu', 'Opteron_G4-v1-x86_64-cpu', 'Opteron_G4-x86_64-cpu', 'Opteron_G5-v1-x86_64-cpu', 'Opteron_G5-x86_64-cpu', 'Penryn-v1-x86_64-cpu', 'Penryn-x86_64-cpu', 'pentium-v1-x86_64-cpu', 'pentium-x86_64-cpu', 'pentium2-v1-x86_64-cpu', 'pentium2-x86_64-cpu', 'pentium3-v1-x86_64-cpu', 'pentium3-x86_64-cpu', 'phenom-v1-x86_64-cpu', 'phenom-x86_64-cpu', 'qemu32-v1-x86_64-cpu', 'qemu32-x86_64-cpu', 'qemu64-v1-x86_64-cpu', 'qemu64-x86_64-cpu', 'SandyBridge-IBRS-x86_64-cpu', 'SandyBridge-v1-x86_64-cpu', 'SandyBridge-v2-x86_64-cpu', 'SandyBridge-x86_64-cpu', 'Skylake-Client-IBRS-x86_64-cpu', 'Skylake-Client-noTSX-IBRS-x86_64-cpu', 'Skylake-Client-v1-x86_64-cpu', 'Skylake-Client-v2-x86_64-cpu', 'Skylake-Client-v3-x86_64-cpu', 'Skylake-Client-x86_64-cpu', 'Skylake-Server-IBRS-x86_64-cpu', 'Skylake-Server-noTSX-IBRS-x86_64-cpu', 'Skylake-Server-v1-x86_64-cpu', 'Skylake-Server-v2-x86_64-cpu', 'Skylake-Server-v3-x86_64-cpu', 'Skylake-Server-x86_64-cpu', 'Snowridge-v1-x86_64-cpu', 'Snowridge-v2-x86_64-cpu', 'Snowridge-x86_64-cpu', 'Westmere-IBRS-x86_64-cpu', 'Westmere-v1-x86_64-cpu', 'Westmere-v2-x86_64-cpu', 'Westmere-x86_64-cpu'] 

```

File: virttest/env_process.py [hunk 1]
```diff
@@ -1027,11 +1027,12 @@ def preprocess(test, params, env):
             env["cpu_model"] = cpu.get_qemu_best_cpu_model(params)
         params["cpu_model"] = env.get("cpu_model")
 
-    if params.get("cpu_driver") is None:
+    qemu_path = utils_misc.get_qemu_binary(params)
```
Comment (luckyh) [inline virttest/env_process.py]: Suggest:
```python
if vm_type == "qemu":
    qemu_path = utils_misc.get_qemu_binary(params)
    if (utils_qemu.has_device_category(qemu_path, "CPU")
        ...
```

Comment (PaulYuuu) [reply virttest/env_process.py]: Updated

**Comment by luckyh** [comment]
Thanks, let's merge this patch first since it's quite important.

**Pull request merged**